asset/asset.go

```diff
@@ -944,6 +944,364 @@ type GroupKeyReveal interface {
+// following structure:
+//
+//	           [tapscript root]
+//	               /       \
+//	[genesis asset ID]   [internal key hash]
+//
+// Here, the tapscript tree consists of two leaf nodes: the genesis asset ID
+// and the hash of the taproot internal key. In this case, we can prove that
+// the genesis asset ID is committed to the tree root by deriving the tapscript
+// root hash using the hash of the internal key. Further, the internal key hash
+// is not a valid asset ID because its pre-image is 33 bytes long (the
+// serialized compressed public key), making it cryptographically improbable
+// to serve as a valid pre-image for an asset ID (which requires a longer
+// pre-image).
+//
+// When a custom tapscript tree is provided, the tapscript tree adopts the
+// following structure:
+//
+//	           [tapscript root]
+//	               /       \
+//	[genesis asset ID]   [tweaked custom tree root]
+//	                            /        \
+//	             [internal key hash]   [custom tree root]
+//
+// In this scenario, we can prove that the root hash commits to the genesis
+// asset ID by deriving the tapscript root from the tweaked custom tree root
+// and the genesis asset ID. Specifically, the tapscript root hash is
+// reconstructed by concatenating the genesis asset ID and the tweaked custom
+// tree root, and then hashing them.
+//
+// Additionally, we can prove that the tweaked custom tree root is not a valid
+// asset ID because it is derived from the internal key hash and the custom
+// tree root. By including the internal key hash as a leaf, the structure
+// ensures that the tweaked custom tree root cannot be misinterpreted as a
+// valid genesis asset ID.
+//
+// Note that the scripts in the tapscript tree will be made non-executable using
+// OP_RETURN, except for the custom tree root.
+type GroupKeyRevealTapscript struct {
+	// finalRoot is the final tapscript root after all tapscript tweaks have
+	// been applied. The asset group key is derived from this root and the
+	// internal key.
+	finalRoot chainhash.Hash
+
+	// customTapscriptRoot is the optional root of a custom tapscript tree
+	// that includes script spend conditions for the group key.
+	customTapscriptRoot fn.Option[chainhash.Hash]
+}
+
+// Validate checks that the group key reveal tapscript is well-formed and
+// compliant.
+func (g *GroupKeyRevealTapscript) Validate(assetID ID,
+	internalKey btcec.PublicKey) error {
+
+	var zeroHashBytes chainhash.Hash
+
+	// Ensure that the final root is not all zeros.
+	if g.finalRoot == zeroHashBytes {
+		return fmt.Errorf("group key reveal final tapscript root is " +
+			"empty")
+	}
+
+	// Extract custom tapscript subtree root if specified. This can be a
+	// hash of a leaf or branch node. We use an empty hash (all zero byte
+	// array) if the custom tapscript root is not specified.
+	customTapscriptRoot := g.customTapscriptRoot.UnwrapOr(chainhash.Hash{})
```

Comment:
> Should we really be inserting a "fake" root here? IMO it would feel more natural to just promote the internal key leaf to the same level as the asset ID leaf instead of inserting a fake, all-zero branch.

Using all zero as the custom tapscript root if absent has no cost to the user, because there isn't a script spend path at that point. My goal with that approach was to have a consistent tree topology whether a custom branch was specified or not. Which means there is only one tapscript root to try to reproduce during verification.

> Either way, I think this approach requires a bit more explanation in a BIP than the "there's a sibling you need to add to the root of your tree" approach I was suggesting.

This means that the user has to modify their tree before tapd can accept it. In my opinion, it would be simpler for the user if they didn’t have to modify their tree at all. I think Tapd should be able to accept any user branch as-is and then provide a control block proof in return.

Reply:
I really don't think tapd needs to do nor should do any of the following:

* user script validation (i.e. user single script leaf can't be asset ID)
* user subtree validation (i.e. user subtree must include sibling node)
* multiple possible validation paths (i.e. final tapscript tree could have different topologies)
* or even produce a user control block proof that can vary in size.

They are all just unnecessary UX problems and potential errors IMO.

Reply:
We can likely simplify the tapscript tree by removing the internal key leaf and using the following structure:
```
//	           [tapscript root]
//	               /       \
//	[genesis asset ID]    [tweaked user subtree]
//	                            /        \
//	                   [OP_RETURN 0]   [user tapscript subtree]
```

A spending script cannot be derived for `[tweaked user subtree]`. Moreover, it is computationally infeasible for a user to construct `[user tapscript subtree]` in a way that would make `[tweaked user subtree]` include a valid asset ID leaf script.

Reply:
We can likely simplify the tapscript tree by removing the internal key leaf and using the following structure:
```
//	           [tapscript root]
//	               /       \
//	[genesis asset ID]    [tweaked user subtree]
//	                            /        \
//	                   [OP_RETURN 0]   [user subtree]
```

A spending script cannot be derived for `[tweaked user subtree]`. Moreover, it is computationally infeasible for a user to construct `[user tapscript subtree]` in a way that would make `[tweaked user subtree]` a valid asset ID leaf.

I think this formulation might be simpler to document, have the same security properties, and be easier to implement.